### ISSUE TITLE
Race Condition with rendering parts in goroutines with templ-initialized context

### ISSUE THREAD

**Issue by kanstantsin-chernik** [a-h/templ#977]
In the current implementation, there is an assumption that all the parts are rendered in a single thread. Context value for children is modified for the whole context. If snippets need to be rendered in goroutines, it creates a race condition where children go missing or sporadically added

**To Reproduce**
https://github.com/kanstantsin-chernik/templ/blob/suspense-async-rendered/examples/suspense-async-rendered/main.templ

I also added comments to the generated code

**Expected behavior**
Each templ component is independent and can be rendered in a goroutine

**Screenshots**
<img width="281" alt="Screenshot 2024-11-01 at 1 31 38 PM" src="https://github.com/user-attachments/assets/29e2fe1e-a7da-45c3-9909-426ae503b7bd">

**Logs**
```
WARNING: DATA RACE
Read at 0x00c01e4a13f0 by goroutine 1053:
  github.com/a-h/templ.GetChildren()
      external/com_github_a_h_templ/runtime.go:78 +0x186

Previous write at 0x00c01e4a13f0 by goroutine 1054:
  github.com/a-h/templ.ClearChildren()
      external/com_github_a_h_templ/runtime.go:68 +0x44
```

**`templ info` output**
 % templ info
(✓) os [ goos=linux goarch=amd64 ]
(✓) go [ location=/home/user/go-code/bin/go version=go version go1.23.2 X:nocoverageredesign linux/amd64 ]
(✓) gopls [ location=/opt/go/path/bin/gopls version=golang.org/x/tools/gopls v0.14.2
    golang.org/x/tools/gopls@v0.14.2 h1:sIw6vjZiuQ9S7s0auUUkHlWgsCkKZFWDHmrge8LYsnc= ]
(✓) templ [ location=/home/user/go-code/bin/templ version=v0.2.778 ]

**Desktop (please complete the following information):**
 - OS: all
 - templ CLI version (`templ version`):  v0.2.778
- Go version (`go version`): go version go1.23.2 X:nocoverageredesign linux/amd64
- `gopls` version (`gopls version`):golang.org/x/tools/gopls v0.14.2

**Additional context**

**Comment by kanstantsin-chernik** [comment]
A couple of questions:
* Should Templ support this example by not reusing context or sync rendering is by design?
* If it is by design, can we at least introduce something like WithoutContext to reset it before we launch go routines?

**Comment by a-h** [comment]
Hi, just looking into this. Wouldn't you also want to put a mutex around the `io.Writer` that you're writing to, in order to avoid the danger of multiple components rendering to the writer at the same time?

If you did that, I don't think you would have a race condition. The suspense example in the repo uses a `for` loop over a channel which means that although data is fetched in parallel, the rendering to the response is sequential, which is typically what you want to prevent multiple writers writing data at the same time.

So, I don't really understand what you're trying to do. Are you thinking that you could get better performance by rendering multiple parts of a page in parallel, or something else?

If it's that you think you'll get better performance by rendering individual sections, then I think you could avoid the race condition by having some HTTP middleware that initializes the context as it receives a request, to enable that, but I suspect that the overhead of creating buffers and copying them around would negate any benefits.

**Comment by a-h** [comment]
So, the objective is to render various sections to sub-buffers, then to render them out in the expected sequence.

I can see how that could, in some cases, improve the performance.

However, @joerdav's comments on the PR are correct. The context design currently expects single threaded use, and `Once`, script rendering and others would require some sort of mutex on the existing context value to function correctly.

For example:

https://github.com/a-h/templ/blob/8e2db03531d9f905c69136dac1b64ee99fc9aa0f/runtime.go#L502C1-L515C2

It expects that it can read and write to the context without a mutex.

https://github.com/a-h/templ/blob/8e2db03531d9f905c69136dac1b64ee99fc9aa0f/once.go#L53-L62

We could consider updating the way that the context works so that children are separate.

This would allow us to have something like this:

```templ
package main

import (
	"bytes"
	"context"
	"errors"
	"io"
	"os"
	"sync"

	"github.com/a-h/templ"
)

func main() {
	// Create a new component.
	page := Page()

	// Render the component.
	page.Render(context.Background(), os.Stdout)
}

func Concurrent(components ...templ.Component) templ.Component {
	return ConcurrentComponent{Components: components}
}

type ConcurrentComponent struct {
	Components []templ.Component
}

func (c ConcurrentComponent) Render(ctx context.Context, w io.Writer) error {
	var wg sync.WaitGroup
	wg.Add(len(c.Components))

	// When all components are done, send a signal.
	done := make(chan struct{})
	go func() {
		wg.Wait()
		done <- struct{}{}
		close(done)
	}()

	buffers := make([]*bytes.Buffer, len(c.Components))
	contexts := make([]context.Context, len(c.Components))
	cancels := make([]context.CancelFunc, len(c.Components))
	for i := 0; i < len(buffers); i++ {
		buffers[i] = new(bytes.Buffer)
		contexts[i], cancels[i] = context.WithCancel(ctx)
	}
	cancel := func() {
		for i := range cancels {
			cancels[i]()
		}
	}
	errs := make([]error, len(c.Components))

	for i := range c.Components {
		go func(i int) {
			defer wg.Done()
			errs[i] = c.Components[i].Render(contexts[i], buffers[i])
		}(i)
	}
	select {
	case <-ctx.Done():
		// If the parent context is cancelled, cancel all the child contexts.
		cancel()
		return ctx.Err()
	case <-done:
		if err := errors.Join(errs...); err != nil {
			return err
		}
	}

	// Render the results.
	for i := range buffers {
		if _, err := io.Copy(w, buffers[i]); err != nil {
			return err
		}
	}
	return nil
}
```

And use it like this:

```templ
templ Page() {
	<!DOCTYPE html>
	<html>
		<head>
			<title>Page</title>
		</head>
		<body>
			<h1>Page</h1>
			@Concurrent(ExpensiveChild(1), ExpensiveChild(2), ExpensiveChild(3), ExpensiveChild(4), ExpensiveChild(5), ExpensiveChild(6), ExpensiveChild(7), ExpensiveChild(8), ExpensiveChild(9), ExpensiveChild(10))
		</body>
	</html>
}

templ ExpensiveChild(i int) {
	@templ.Flush() {
		<div>
			<h2>Expensive Child { strconv.Itoa(i) }</h2>
			@ExpensiveComponent()
		</div>
	}
}

func ExpensiveComponent() templ.Component {
	return templ.ComponentFunc(func(ctx context.Context, w io.Writer) error {
		time.Sleep(1 * time.Second)
		_, err := w.Write([]byte("Expensive Component"))
		return err
	})
}
```

With the expected output.

```html
<!doctype html><html><head><title>Page</title></head><body><h1>Page</h1><div><h2>Expensive Child 1</h2>Expensive Component</div><div><h2>Expensive Child 2</h2>Expensive Component</div><div><h2>Expensive Child 3</h2>Expensive Component</div><div><h2>Expensive Child 4</h2>Expensive Component</div><div><h2>Expensive Child 5</h2>Expensive Component</div><div><h2>Expensive Child 6</h2>Expensive Component</div><div><h2>Expensive Child 7</h2>Expensive Component</div><div><h2>Expensive Child 8</h2>Expensive Component</div><div><h2>Expensive Child 9</h2>Expensive Component</div><div><h2>Expensive Child 10</h2>Expensive Component</div></body></html>
```

Really, we should also adjust the children behaviour. At the moment, you get one child component, but it would probably be nicer syntax to do:

```templ
@Concurrent() {
  <div>Level 1</div>
  <div>Level 2</div>
}
```

i.e. we return a slice of components (as a type that implements templ.Component).

I've added a draft PR above.

**Comment by joerdav** [comment]
It all looks good to me. I'd expect the performance implications to be minimal and only really come into play in these very niche scenarios where concurrent renders are using the same ctx.

A couple of things come to mind, on a bit of a tangent, I do think that children being a `[]Component` is a useful idea...

- I wonder if then the `@` syntax should be able to render a slice of components?
- Is then the `{ children... }` syntax unneeded? We can just do `@templ.GetChildren(ctx)`

In general though, the addition of the mutex, I can't see any issues here besides one.. this could mean that the rendered result is not going to always be consistend. Imagine you have 2 components Foo and Bar, both render a component which contains a `Once`, our page renders Foo above Bar but does so concurrently. Templ will correctly only render our `Once` one time, however there is a bit of a race, will the Once render in Foo or Bar, it's undetermined. Maybe this is okay?

**Comment by a-h** [comment]
Yes, I think the `{ children... }` syntax could be retired, which would be a good thing, because it's one less piece of magic. An update to `templ fmt` could remove it from files where it exists.

I thought `templ.GetChildren` could return a `type ComponentSlice []Component` which would implement `templ.Component` by having a `Render` function, but your idea would work too.

With `ComponentSlice`, it's still a `templ.Component`, so there would be no breaking changes, but it would be possible to inspect the type, e.g. `components, isComponentSlice := c.(templ.ComponentSlice)` if specific handling of the children was wanted.

Good point on the race condition for `Once`, I hadn't thought of that! Similarly, there will be cases where a script needs to be rendered to the output _before_ the HTML, and if happens out of order, that would mess with it...

A concurrent renderer would need to take that into account somehow.

**Comment by whalenrp** [comment]
Thanks for the suggestion. Let me pull out the relevant bits to restate:

```
templ StreamingPage(mySeq iter.Seq[func() string]) {
    ...
    for data := range mySeq {
        @await() {
            <div>{ data() }</div>
        }
    }
}
```

I do have something similar where each call to at iterator returns markup -- however I think what you've got here will discard the context. If you're using templ to render each child it won't support context lookups (i.e. [prop drilling](https://templ.guide/syntax-and-usage/context/#prop-drilling)) for any custom values, and [templ.Once](https://templ.guide/syntax-and-usage/render-once/) won't work correctly for the overall page. Since templ's internal context state is just one of many values propagating through the `context`, we're doing some manipulation to make sure each `data()` call has access to those values and overriding the templ state in each goroutine to avoid concurrency issue. Something along the lines of:

```
func newWrapped(valuesContext context.Context) context.Context {
	return &wrappedCtx{
		Context: templ.InitializeContext(context.Background()), // allocates maps for every child component
		values:  valuesContext,
	}
}

func (c *wrappedCtx) Value(key any) any {
	if val := c.Context.Value(key); val != nil { // templ state lookups only hit this block
		return val
	}
	return c.values.Value(key) // everything else traverses the original context
}
```

However, this breaks the Once behavior, and supporting this inside templ itself seems like the cleanest way to fix it.

**Comment by scrhartley** [comment]
Hey @whalenrp.
With my adoption of the Future's as functions pattern, each one represents a value, not a component. I'm just introducing some laziness, with the concurrency hidden under the covers where Templ can't see it. There is some potential for confusion since `ConcurrentSeq` in my repo is not the same as the one in yours. My one is for when you specifically aren't interested in maintaining order. If you want in-order you can just create a bunch of independent futures and access them in the desired order in the template. If they are related, perhaps: `[]func() string` or potentially there's a case where you'd need `iter.Seq[func() string]`, but to me that's a dynamic-data concern, not a template concern.

With my approach if you wanted different guarantees, nested Futures could also be used:
```go
type compoundFuture struct {
	StaticValue string
	LazyValue1  string
	LazyValue2  string
}
compoundValue := templates.Concurrent(func() compoundFuture {
	return compoundFuture{"Hello", myFuture1(), myFuture2()}
})
```
(So `compoundValue` is a Future which is only ready when everything inside is ready.)

**Comment by whalenrp** [comment]
Thanks for your response!

I'm happy to update the buffers and can do that before submitting that part of the stack for review.

Regarding the benchmarks, I (with some help) put up some numbers and the benchmark test that was used for them [here](https://github.com/whalenrp/templ/blob/benchmark/syncmap-overhead-analysis/syncmap_benchmark_analysis.md). The TL;DR is that the deduping features, if used, contribute heavily to overall performance numbers. If you don't use any of those features then there's clearly no impact to using `sync.Map` by default, but if you DO rely on those that logic drives a majority of the templ rendering overhead. 

<img width="489" height="256" alt="Image" src="https://github.com/user-attachments/assets/30284835-39ca-4c2f-859e-c29caf6de967" />

Ultimately though, the numbers from templ itself are still overall small and they only seem large since it's a maximum 2x penalty on a small baseline number. I just wasn't sure how you guys would weigh performance vs code cleanliness in a case like this and opted for better performance for the default case (single-threaded rendering)

**Comment by whalenrp** [comment]
I think we can close this one out now, right? Or do you guys want to leave it open to track whether the API may change to support a component that does this concurrent rendering natively like described [here](https://github.com/a-h/templ/issues/977#issuecomment-2567066919)?

**Comment by a-h** [comment]
Yeah, thanks, I think this is good to close. We can always revisit a more specific requirement and handle it separately.

**Comment by a-h** [comment]
Thanks for your focus on this issue.
